server/routes/companies.js

```diff
@@ -116,25 +179,399 @@ router.get('/filter', function(req, res, next) {
   const mid = req.query.mid;
 
   if (pid && mid) {
-    db.query('SELECT DISTINCT company_major.company_id FROM company_major LEFT OUTER JOIN company_position ON (company_major.company_id = company_position.company_id) WHERE (position_id = $1 AND major_id = $2);', [pid, mid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex.distinct('company_major.company_id').from('company_major')
+    .leftOuterJoin('company_position', 'company_major.company_id', 'company_position.company_id')
+    .where({position_id: req.body.pid, major_id: req.query.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (mid) {
-    db.query('SELECT company_id FROM company_major WHERE major_id = $1;', [mid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_major').select('company_id').where({major_id: req.body.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (pid) {
-    db.query('SELECT company_id FROM company_position WHERE position_id = $1;', [pid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_position').select('company_id').where({position_id: req.body.pid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else { 
-    return res.status(400).send({
-      message: '\'mid\' and \'pid\' are undefined'
-    });
+    util.throwError(400, '\'mid\' and \'pid\' are undefined');
   }
 });
 
+// Add a single company or related data
+router.post('/', function(req, res, next) {
+  values = {
+    name: req.body.name,
+    website: req.body.website,
+    logo: req.body.logo,
+    citizenship_requirement: req.body.citizenship_requirement,
+    description: req.body.description
+  }
+  let ranks = req.body.rank;
+  let user_ids = req.body.user_id;
+  let position_ids = req.body.position_id;
+  let major_ids = req.body.major_id;
+  let contact_ids = req.body.contact_id;
+  let event_ids = req.body.event_id;
+
+  console.log(req.body.position_id);
+
+  var queryCompany = knex('company').insert(values)
+  knex.transaction(function(trx) {
+    return queryCompany.transacting(trx)
+      .returning('id')
+      .then(async function(ids){
+        let company_id = ids[0];
+
+        if (ranks && user_ids) {
```

Comment:
Create the queries/arrays/all this outside of the transaction. This will simplify the code and make it much more readable

Reply:
We might have issues with doing this in the post request since the company_id can only be accessed after the first insertion, in the transaction. I think the best we can simplify this is to use a for loop outside the transaction for each value except company id, and then loop inside adding the company id to each element of the array (eg. for positions, put all position ids outside the transaction in an array of JS objects, then add a company Id attribute for each object inside the transaction, since we must insert the company before being able to access the id). If this is the case, I’m not sure if we should do this since it splits the code into two for loops (one inside, one outside). What do you think?

This is definitely implementable for the put method!

Reply:
Ah I definitely see your point. My main hangup is that it's hard to read rn with so many nested loops/checks/callbacks/functions.

I think it'll be best to do the 2 for loops so we can do error checks outside of the transaction and simplify the transaction more. With our scope, performance will have no noticeable effects

Reply:
Also loop iteration has some nifty functions in Javascript that should make adding the company_id very simple/compact

Reply:
Note: I pushed some changes that fixed part of the put method, still working on extracting remove statements

Reply:
Finished cleaning up the put request. We can work on the 2 loops